src/Symfony/Component/Messenger/Failure/FailedMessageHandler.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Failure;
+
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\Messenger\Exception\LogicException;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Messenger\Stamp\DelayStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+use Symfony\Component\Messenger\Stamp\RedeliveryStamp;
+use Symfony\Component\Messenger\Stamp\SentStamp;
+
+/**
+ * Handles a FailedMessage by redelivering via the original transport.
+ *
+ * @author Ryan Weaver <ryan@symfonycasts.com>
+ *
+ * @experimental in 4.3
+ */
+class FailedMessageHandler
+{
+    private $messageBus;
+    private $logger;
+
+    public function __construct(MessageBusInterface $messageBus, LoggerInterface $logger = null)
+    {
+        $this->messageBus = $messageBus;
+        $this->logger = $logger ?: new NullLogger();
+    }
+
+    public function __invoke(FailedMessage $failedMessage)
+    {
+        $failedEnvelope = $failedMessage->getFailedEnvelope();
+        $envelope = $failedMessage->getFailedEnvelope()
+            ->withoutAll(DelayStamp::class)
+            ->withoutAll(RedeliveryStamp::class)
+            ->withoutAll(ReceivedStamp::class);
+
+        if ($failedMessage->isStrategyRetry()) {
+            // ReceivedStamp is needed to fake that the message was received in the original way
+            $receivedStamp = $failedEnvelope->last(ReceivedStamp::class);
+            if (null === $receivedStamp) {
+                throw new LogicException('Cannot retry failed message: the original envelope lacks a ReceivedStamp.');
```

Comment:
shouldn't this use `UnrecoverableMessageHandlingException` as it would otherwise retry the retry ;) ?